hypothesis-python/src/hypothesis/extra/ghostwriter.py

```diff
@@ -785,8 +785,88 @@ def _get_qualname(obj: Any, *, include_module: bool = False) -> str:
     return qname
 
 
+class BaseNameVisitor(ast.NodeVisitor):
+    def __init__(self):
+        self.base_name = None
+
+    def visit_Attribute(self, node: ast.Attribute):
+        self.visit(node.value)
+
+    def visit_Subscript(self, node: ast.Subscript):
+        self.visit(node.value)
+
+    def visit_Call(self, node: ast.Call):
+        self.visit(node.func)
+
+    def visit_Name(self, node: ast.Name):
+        assert self.base_name is None
+        self.base_name = node.id
+
+def potentially_modified_vars(func: ast.FunctionDef, params: Set[str]) -> Set[str]:
+    """
+    We want to get any variables that might be modified in a way that might
+    cause side-effects so that we can copy them.
+
+    A variable are potentially modified if:
+    - a method is called on it, an attribute under it, or an index
+      of it is anywhere on the LHS of an ast.Assign or an ast.AugAssign
+    - a method called on it, an attribute of it, or an index of it 
+        - Do we want to count methods on the RHS?
```

Comment:
I'd only count method calls if they're top-level statements, not part of an assignment or inside an index or whatever - the heuristic being that methods typically either mutate state, xor have a return value.

```python
arg[index] = ...
arg.attr = ...
arg.method()

result = arg.other_method()  # probably safe
```

Reply:
I'm not entirely convinced that this heuristic works-- I think it is a safe assumption that a standard accessor doesn't modify anything (which is not necessarily true, but is close enough to true) but the pattern of a method that mutates and then returns is one that's relatively common-- for instance, even in the standard library, [`IO.read` returns a value and modifies the offset](https://github.com/python/cpython/blob/3.12/Lib/typing.py#L3205). 

Files are, of course, a case where side effects exist outside of the program where copying wouldn't help, but the pattern itself is very well established in Python, and this was a simple example!